Try: Always full-height code editor

jasmussen opened the pull request:
This PR changes the code editor to not use the autosizing textarea component, and instead to scroll independently and be always full-height.

<img width="1232" alt="Screenshot 2020-04-28 at 15 54 52" src="https://user-images.githubusercontent.com/1204802/80496230-31e6c480-8969-11ea-8f2e-c1b9fecdfb55.png">

Part of the reason for this was that the autosizing textarea was potentially slow in Safari. Worth testing and comparing with this branch.

jasmussen committed:
Try: Always full-height code editor

This PR changes the code editor to not use the autosizing textarea component, and instead to scroll independently and be always full-height.

youknowriad commented on packages/edit-post/src/components/text-editor/style.scss:
```diff
@@ -1,6 +1,7 @@
 .edit-post-text-editor {
 	position: relative;
 	width: 100%;
+	height: calc(100% - 56px);
```
Where this 56px comes from? can we avoid it and use flex growing?

youknowriad reviewed (approved):
Seems to work well, I guess I'd have preferred a borderless editor completely 100% height and width maybe. But it seems like a too drastic change for now.

youknowriad commented:
I think there might be some snapshots to change or something like that for the JS unit tests.

jasmussen committed:
Fix tests.

jasmussen commented:
Okay, I addressed the feedback. The code editor now uses flex to size things. It's also mobile optimized.

However, it's also now a semi-big change: it drops the react autosizing textarea in favor of a plain textarea, necessitating the tests to be updated (I would appreciate a sanity check on that code). So there's actually a lot of layout changing in this PR.

For that reason, I gave it a good test. Chrome emulator:

<img width="444" alt="chrome emulator" src="https://user-images.githubusercontent.com/1204802/80575961-a6bb0c80-8a04-11ea-844b-c5b247fb6a67.png">

Chrome small:

<img width="1289" alt="chrome small" src="https://user-images.githubusercontent.com/1204802/80575970-a9b5fd00-8a04-11ea-8b89-3ff1efe0a479.png">

Chrome big;

<img width="1920" alt="chrome big" src="https://user-images.githubusercontent.com/1204802/80575979-ad498400-8a04-11ea-916d-45f7b1c76581.png">

Safari:

<img width="1194" alt="safari" src="https://user-images.githubusercontent.com/1204802/80575989-b175a180-8a04-11ea-9be0-f2b2ca5e9a33.png">

iPhone simulator:

![iphone](https://user-images.githubusercontent.com/1204802/80576007-b89caf80-8a04-11ea-9060-5f3917329932.gif)

**Hold up** — you can type off the view of the screen, there! That seems bad! 

It does, it deserves another PR. I believe some of the fixed position stuff that we do for the editor _exactly_ to improve this, has not been ported sufficiently to the text editor. 

When I saw this initially, I thought this might be too big a change to ship in 8.0. However then I compared to what's currently shipping:

![previous version](https://user-images.githubusercontent.com/1204802/80576103-e71a8a80-8a04-11ea-8250-a766fc9d55c1.gif)

It appears that react autosize for the entire editable area plays badly with mobile Safari and the positioning shenanigans the softkeyboard does to the viewport. As you can see, it's barely usable. In other words, this PR is likely to be a massive step up from that, in that it removes the autosizing text field in favor of a plain textfield. 

But it does suggest this PR needs another sanity check, if you have time.

youknowriad commented on packages/editor/src/components/post-text-editor/style.scss:
```diff
@@ -1,28 +1,35 @@
-	width: 100%;
-	box-shadow: none;
-	resize: none;
+.edit-post-text-editor__body {
```
Should the styles for `.edit-post-text-editor__body`  be moved to ` packages/edit-post/src/components/text-editor/style.scss`  instead?

youknowriad replied:
(not the textarea itself but its wrapper)

youknowriad reviewed (commented):
The code changes look good to me. I didn't test extensively though.

pwkip commented:
I tested in ubuntu with chrome. Most of it looks fine. Maybe just add a min-height to the textarea in case the title gets too long?

![Screenshot from 2020-04-29 13-03-46](https://user-images.githubusercontent.com/1150196/80589372-551d7c80-8a1a-11ea-9327-849d60df4341.png)

jasmussen commented:
Thank you for the reviews. Some excellent points have been brought up. Because this is an important bit, I'll give this some more time in the oven, address feedback and polish it up. So it will likely miss the 8.0 deadline.

jasmussen committed:
Add scrollfix.

jasmussen commented:
The iPhone behavior frustrated me, so I dug a bit and found a problem in https://github.com/WordPress/gutenberg/pull/21950/commits/cdedfb1da2547468e256ccfa4d0f9f81dd2bc909.

When applied to master, it looks like this:

![master with scrollfix](https://user-images.githubusercontent.com/1204802/80795099-dc462e00-8b9b-11ea-8936-45ca3bacfc9a.gif)

When applied to this branch, it looks like this:

![fullheight with scrollfix](https://user-images.githubusercontent.com/1204802/80795109-e0724b80-8b9b-11ea-918e-8ac8b396bf8a.gif)

Both are somewhat bad, but they are also both usable, especially compared to [what's shipping](https://github.com/WordPress/gutenberg/pull/21950#issuecomment-621065738), and at least they are both behaving similarly.

jasmussen committed:
Add max-height to title.

jasmussen commented:
> Maybe better to check for visual editor if we ever add a new mode?

Good point. I'm actually working in a separate branch, because I'm starting to believe this refactor isn't the best way to do it. More to come in a minute.

jasmussen commented:
I created https://github.com/WordPress/gutenberg/pull/22021 as an alternative to this, and must admit I prefer it.

jasmussen commented:
Even though this one was approved, for now I'm going to close it in favor of #22021 which is better in most ways. Would very much appreciate a review.
